### PR TITLE
list parameter

### PR DESCRIPTION
**Pull request by gshiba**
Hello!

This PR handles #363 

```
papermill input.ipynb output.ipynb -p my_list abc def ghi
```

But, it arguably relies on hacking `click.Option`, and `click` authors seem to be against variadic options [1].

Full credit goes to user stephen-rauch on stackoverflow [2]

The `class OptionEatAll` should probably be placed somewhere better...

[1] https://github.com/pallets/click/pull/259
[2] https://stackoverflow.com/a/48394004

### CODE REVIEW

**Comment by MSeal** [comment]
Thanks for the PR! This change seems solid, despite the nasty click hacks required. Let me fix the failing test (it's failing on master too) and make play around with this change some before merging.

**Comment by gshiba** [comment]
The `-r/--raw` option should probably get the same treatment?

File: papermill/cli.py [hunk 1]
```diff
@@ -171,10 +217,20 @@ def papermill(
         parameters_final.update(read_yaml_file(files))
     for params in parameters_yaml or []:
         parameters_final.update(yaml.load(params, Loader=NoDatesSafeLoader))
-    for name, value in parameters or []:
-        parameters_final[name] = _resolve_type(value)
-    for name, value in parameters_raw or []:
-        parameters_final[name] = value
+    for name, *values in parameters or []:
```
Comment (MSeal) [inline papermill/cli.py]: This isn't legal syntax in python 2 (which we're still supporting through the end of the year)

**Comment by gshiba** [comment]
Thanks for reviewing!

I hopefully fixed the py2 compatibility issues -- tests are still failing though.

Ok with where I placed the `class OptionEatAll`?

File: papermill/cli.py [hunk 2]
```diff
@@ -171,10 +217,22 @@ def papermill(
         parameters_final.update(read_yaml_file(files))
     for params in parameters_yaml or []:
         parameters_final.update(yaml.load(params, Loader=NoDatesSafeLoader))
-    for name, value in parameters or []:
-        parameters_final[name] = _resolve_type(value)
-    for name, value in parameters_raw or []:
-        parameters_final[name] = value
+    for name_values in parameters or []:
+        name, values = name_values.pop(0), name_values
```
Comment (MSeal) [inline papermill/cli.py]: I _think_ this is causing the test failures.
```suggestion
        name, values = name_values[0], name_values[1:]
```

File: papermill/cli.py [hunk 3]
```diff
@@ -171,10 +217,22 @@ def papermill(
...
+        if not values:
+            raise ValueError("There must be at least one value for key '{}'".format(name))
+        elif len(values) == 1:
+            parameters_final[name] = _resolve_type(values[0])
+        else:
+            parameters_final[name] = [_resolve_type(v) for v in values]
+    for name_values in parameters_raw or []:
+        name, values = name_values.pop(0), name_values
```
Comment (MSeal) [inline papermill/cli.py]:
```suggestion
        name, values = name_values[0], name_values[1:]
```

File: papermill/cli.py [hunk 4]
```diff
@@ -171,10 +217,22 @@ def papermill(
...
+        if not values:
+            raise ValueError("There must be at least one value for key '{}'".format(name))
+        elif len(values) == 1:
+            parameters_final[name] = values[0]
+        else:
+            parameters_final[name] = values
```
Comment (MSeal) [inline papermill/cli.py]:
```suggestion
            parameters_final[name] = list(values)
```

**Comment by MSeal** [comment]
I'm good with where you placed OptionEatAll as it's not so large it needs its own file and relates to the problem immediately below it.

I think with the suggested change this will pass and I can add it to the release about to go out.

File: papermill/cli.py (hunk 2 above)
Comment (gshiba) [reply papermill/cli.py]: Hmmm, interesting that they may be the cause.

`name_values[1:]` could raise `IndexError`...

Comment (MSeal) [reply papermill/cli.py]: Nope it's safe:

```
In [1]: t = []                                                                                                                              
In [2]: t[1:]                                                                                                                               
Out[2]: []
```

Comment (gshiba) [reply papermill/cli.py]: Ah.. thanks.

Comment (gshiba) [reply papermill/cli.py]: Ah, I get it now. `name_values` is a `tuple`, not a `list`.

**Comment by MSeal** [comment]
Thanks for the improvement and working through the test failures!

**Pull request merged**